apps/cli/cli_common.c

```diff
@@ -774,6 +774,94 @@ cli_set_mode(clixon_handle h,
     return retval;
 }
 
+/*!
+ * @brief Spawns a new process and runs the specified program within it
+ *
+ * The function creates a new child process and runs the specified program within it.
+ * Before launching the program in the child process, environment variables taken from cvv are set.
+ *
+ * The function checks the parameters passed to it. Situations where at least one function argument
+ * is missing, or when two arguments are present in both the function and the cvv vector are considered invalid.
+ *
+ * Example usage:
+ * @code
+ *      run_program_err("Run program"), cli_start_program();
+ *      run_program_python3("Run program"), cli_start_program("python3");
+ *      run_program_python3_source_arg("Run program"), cli_start_program("python3", "/tmp/test.py");
+ *      run_program_python3_source_arg_vector("Run program") <source:rest>("Path program"), cli_start_program("python3");
+ *      run_program_python3_source_arg_vector_err("Run program") <source:rest>("Path program"), cli_start_program("python3", "/tmp/test2.py");
+ *      run_program_bash("Run program"), cli_start_program("bash");
+ * @endcode
+ *
+ * @param[in] h Clixon handle
+ * @param[in] cvv Vector of command variables
+ * @param[in] argv Function arguments
+ * @retval   0 OK, returns the exit code of the program
+ * @retval  -1 Error
+ *
+ */
+int
+cli_start_program(clixon_handle h,
+                      cvec         *cvv,
+                      cvec         *argv)
+{
+    int pid = 0;
+    int retval = -1;
+    char *script_path = NULL;
+    char *runner = NULL;
+    struct passwd *pw = NULL;
+
+    /* Check parameters */
+    if (cvec_len(argv) == 0){
+        clixon_err(OE_PLUGIN, EINVAL, "Can not found argument in a function");
+        goto done;
+    }
+    if ((cvec_len(argv) >= 2) && (cvec_len(cvv) >= 2)){
+        clixon_err(OE_PLUGIN, EINVAL, "A lot of arguments");
+        goto done;
+    }
+    if ((cvec_len(argv) == 2) && (cvec_len(cvv) == 2)){
+        clixon_err(OE_PLUGIN, EINVAL, "You cannot use 2 arguments in a function and 1 argument in a vector");
+        goto done;
+    }
+
+    /* get data */
+    if (cvec_len(argv) == 1){
+        runner = cv_string_get(cvec_i(argv, 0));
+    }
+    if (cvec_len(argv) == 2){
+        runner = cv_string_get(cvec_i(argv, 0));
+        script_path = cv_string_get(cvec_i(argv, 1));
+    }
+    if (cvec_len(cvv) == 2){
+        script_path = cv_string_get(cvec_i(cvv, 1));
+    }
+
+    if ((pw = getpwuid(getuid())) == NULL){
+        clixon_err(OE_PLUGIN, errno, "getpwuid");
+        goto done;
+    }
+    if (chdir(pw->pw_dir) < 0){
```

Comment:
Here you are changing the directory for the whole process. It may impact other services that runs on the same process and rely on the current working directory.
It is better to do the following:
1. fork() the process
2. for the child process:
  a. change the current directory of the child process (it will leave the current directory of the main process unchanged)
  b. execute the script you want
3. for the parent process - wait for the child process to be finished

Other notes: 
1. In case of failure, you execute `exit(0)` - this will terminate the whole process including other services that may run in this process. It will be better to report an error and let the upper layer decide how to handle errors. 
2. It is better to pass the directory from where to execute the `runner` as a separate argument. It will allow the caller to decide what is the current directory from where you run the script. (If the caller does not specify the directory, you can use the initial working directory by default). This is important especially if you plan to use Python scripts that may require to be executed from a specific directory that is different from the initial working directory - this may allow the Python interpreter to find its dependencies.

Reply:
done